### PR TITLE
Multistore publish synchronize

### PR DESCRIPTION
**Pull request by gerner-spryker**
- Developer(s): @ehsanmx, @gerner-spryker

- Peer Reviewer: @ehsanmx 

- Ticket: https://spryker.atlassian.net/browse/CORE-2793
- Ticket: https://spryker.atlassian.net/browse/CORE-2977

- Core PR: https://github.com/spryker/spryker/pull/3132

- Release Overview: -

#### Please confirm

- [x] No new OS components - or they have been approved by legal department
- [ ] All new or heavily changed classes get an "A" rating (Scrutinizer), except Facades, Factories, QueryContainers and Tests

#### Documentation

- [x] Functional documentation provided or in progress?
- [x] Integration guide for projects provided or not needed?
- [x] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   RabbitMq              | major                |  Store,Queue                 |

#### Release Notes

Allowing handling of multiple RabbitMQ connections. Messages can be published to a specific queue pool or either by using a connection assigned store name. 

-----------------------------------------

#### Module RabbitMq

_**Mini-Major:** Added functionality with incompatible API changes_

Def of done (by responsible developer):
- [x] All methods in facade, client and plugin interfaces provide an [API-doc](https://academy.spryker.com/understanding_spryker/definition_api.html).
- [x] New code fits to the architecture rules.
- [x] All new or changed business logic is covered by functional tests (in Zed business layer).
- [x] Migration guide provided.

Improvements

- (internal) Store dependency added to composer.json to allow configuring RabbitMQ queue pools in default configuration.
- (internal) Connection::queueOptionCollection property is replaced.
- (internal) Connection has a QueueConnection transfer property which includes the previously injected ArrayObject holding the QueueOptionCollection. 
- (internal) Fixing return type of QueueEstablishmentHelper::createQueue().
- (internal) Publisher has a ConnectionManager constructor dependency instead of an AMQPChannel to allow to access any required channel instead of one pre-chosen channel.
- (internal) Publisher::sendMessage() sends provided message based on the set StoreNames, queuePools or defaultChannel.
- (internal) Publisher::sendMessages() sends provided messages based on the set StoreNames, queuePools or defaultChannel.
- RabbitMqConfig reads up or provides multiple queue connection configs instead of one.
- RabbitMqConfig::getQueueConnection() is replaced.
- (internal) Client\RabbitMq\RabbitMqFactory has a static connection manager property instead of the static connection. Dependent objects are access the default connection from the connection manager.
- (internal) Client\RabbitMq\RabbitMqFactory::connection is replaced.
- (internal) Client\RabbitMq\RabbitMqFactory::createConnection() is removed.
- (internal) Client\RabbitMq\RabbitMqFactory::createStaticConnection() is removed.
- (internal) Client\RabbitMq\RabbitMqFactory::getQueueConnectionConfig() is removed.
- (internal) Client\RabbitMq\RabbitMqFactory::createAMQPStreamConnection() is removed.
- QueueConnection transfer object has a "name", "isDefaultConnection" and "storeNames" property.
- Shared\RabbitMq\RabbitMqConstants.php is renamed to Shared\RabbitMq\RabbitMqEnv.php.
- Zed/RabbitMq/RabbitMqConfig::getApiExchangesUrl() uses the newly defined RabbitMqConstants::RABBITMQ_API_VIRTUAL_HOST constant instead of the RabbitMqConstants::RABBITMQ_VIRTUAL_HOST.
- Zed/RabbitMq/RabbitMqConfig::getApiQueuesUrl() uses the newly defined RabbitMqConstants::RABBITMQ_API_VIRTUAL_HOST constant instead of the RabbitMqConstants::RABBITMQ_VIRTUAL_HOST.
- Zed/RabbitMq/RabbitMqConfig::getApiUserPermissionsUrl() uses the newly defined RabbitMqConstants::RABBITMQ_API_VIRTUAL_HOST constant instead of the RabbitMqConstants::RABBITMQ_VIRTUAL_HOST.

### CODE REVIEW

File: src/Spryker/Client/RabbitMq/Model/Connection/Connection.php [hunk 1]
```diff
@@ -39,17 +39,17 @@ class Connection implements ConnectionInterface
...
      * @param \PhpAmqpLib\Connection\AMQPStreamConnection $streamConnection
      * @param \Spryker\Client\RabbitMq\Model\Helper\QueueEstablishmentHelperInterface $queueEstablishmentHelper
-     * @param \ArrayObject $queueOptionCollection
+     * @param \Generated\Shared\Transfer\QueueConnectionTransfer $queueConnection
      */
     public function __construct(
         AMQPStreamConnection $streamConnection,
         QueueEstablishmentHelperInterface $queueEstablishmentHelper,
-        ArrayObject $queueOptionCollection
+        QueueConnectionTransfer $queueConnection
```
Comment (aurisl) [inline src/Spryker/Client/RabbitMq/Model/Connection/Connection.php]: `queueConnectionTransfer`

Comment (aurisl) [reply src/Spryker/Client/RabbitMq/Model/Connection/Connection.php]: Btw why use transfer for this? for configuration i would expect something immutable.

File: src/Spryker/Client/RabbitMq/Model/Connection/Connection.php [hunk 2]
```diff
@@ -62,6 +62,30 @@ public function getChannel()
...
+     */
+    public function getStoreNames()
+    {
+        return $this->queueConnectionConfig->getStoreNames();
+    }
+
+    /**
+     * @return bool
+     */
+    public function getIsDefaultConnection()
```
Comment (aurisl) [inline src/Spryker/Client/RabbitMq/Model/Connection/Connection.php]: isDefaultConnection?

File: src/Spryker/Client/RabbitMq/Model/Connection/ConnectionFactory.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function getQueueConnectionConfigs()
+    {
+        return $this->getConfig()->getQueueConnections();
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\QueueConnectionTransfer $queueConnectionConfig
+     *
+     * @return \PhpAmqpLib\Connection\AMQPStreamConnection
```
Comment (aurisl) [inline src/Spryker/Client/RabbitMq/Model/Connection/ConnectionFactory.php]: can you depend on abstract `AbstractConnection` instead?

File: src/Spryker/Client/RabbitMq/Model/Connection/ConnectionManager.php [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+     *
+     * @return \Spryker\Client\RabbitMq\Model\Connection\ConnectionInterface
+     */
+    protected function getConnection(QueueConnectionTransfer $queueConnectionConfig)
+    {
+        return $this->connectionFactory->createConnection($queueConnectionConfig);
+    }
+
+    /**
+     * @param $storeName
```
Comment (aurisl) [inline src/Spryker/Client/RabbitMq/Model/Connection/ConnectionManager.php]: string

**Comment by ehsanmx** [comment]
This will be merged with PR https://github.com/spryker/rabbit-mq/pull/16

**Pull request closed without merging**